web/src/main/java/org/springframework/security/web/csrf/CookieCsrfTokenRepository.java

```diff
@@ -55,14 +53,15 @@
 
 	private String cookieDomain;
 
-	private Optional<SecureRandom> optionalSecureRandom = Optional.empty();
+	private Function<String, CsrfToken> generateTokenProvider = (value) -> new DefaultCsrfToken(this.headerName,
```

Comment:
Rather than passing in `createNewToken()` as an argument to the `Function` this should be a `Producer` and it can create the token itself.

Reply:
This is the same `generateTokenProvider` found in `CookieCsrfTokenRepository`. We could move this logic as a static method on `DefaultCsrfToken` just like the static methods on `XorCsrfToken`.

Reply:
the `loadToken` method for 
`CookieCsrfTokenRepository` required to passed in string token as value and construct the `CsrfToken` instance which is why it shouldn't be a `Producer` that can create token itself.

Reply:
even if it was replaced with something like `Producer<CsrfToken> generator = new DefaultCsrfTokenGenerator(parameterName, headerName);`

it will fail on this test

	@Test
	public void generateTokenCustom() {
		String headerName = "headerName";
		String parameterName = "paramName";
		this.repository.setHeaderName(headerName);
		this.repository.setParameterName(parameterName);

		CsrfToken generateToken = this.repository.generateToken(this.request);

		assertThat(generateToken).isNotNull();
		assertThat(generateToken.getHeaderName()).isEqualTo(headerName);
		assertThat(generateToken.getParameterName()).isEqualTo(parameterName);
		assertThat(generateToken.getToken()).isNotEmpty();
	}

because the `headerName` and `parameterName` was not in sync

it can be solved by setting it together something like :

	public void setParameterName(String parameterName) {
		Assert.notNull(parameterName, "parameterName is not null");
		this.parameterName = parameterName;
		this.generateTokenProvider.setParameterName(parameterName);
	}

however it will caused another issue for another kind of test:

		String headerName = "headerName";
		String parameterName = "paramName";
		this.repository.setHeaderName(headerName);
		this.repository.setParameterName(parameterName);
		Producer prod = new DefaultCsrfTokenGenerator(headerName, parameterName);

		this.repository.setGenerateToken(prod);
		prod.setHeaderName("otherHeader");

		CsrfToken generateToken = this.repository.generateToken(this.request);

		assertThat(generateToken).isNotNull();
		assertThat(generateToken.getHeaderName()).isEqualTo(headerName); // this will fail
		assertThat(generateToken.getParameterName()).isEqualTo(parameterName);
		assertThat(generateToken.getToken()).isNotEmpty();

the above test will fail because the value was not in sync

Reply:
The Function & FunctionalInterface approach ensures that we are passing `this.headerName` and `this.parameterName` as reference so that it will always be in sync 

this is to ensure backwards compatibility for `setParameterName` and `setHeaderName` while depreciating it before removed in the future.

Reply:
Sorry. I think you are right that the default generator needs to be able to access the headerName and parameterName values in the event they are updated. Alternatively, setting the headerName and parameterName need to also update the generator.

Reply:
So is this approach fine ? should we deprecate the `setHeaderName` and `setParameterName` in both `CookieCsrfTokenRepository` and `HttpSessionCsrfTokenRepository` as well?

Reply:
Yes